### ISSUE TITLE
IMAP Authentication failed

### ISSUE THREAD

**Issue by wAikAp** [jstedfast/MailKit#1553]
**Describe the bug**
I have followed [ExchangeOAuth2.md](https://github.com/jstedfast/MailKit/blob/master/ExchangeOAuth2.md) by using the publicClientApplication to get the accessToken. However, the email Microsoft can be logged in and the web browser said  "Authentication complete. You can return to the application. Feel free to close this browser tab.". But a "MailKit.Security.AuthenticationException" IMAP Authentication failed exception is thrown. Did I do something wrong?
![Screenshot 2023-04-19 at 11 59 32 AM](https://user-images.githubusercontent.com/16193686/232963631-cc4ffb2a-ad3e-4812-a428-f0d4ff124392.png)

The testing Microsoft account is an external account and added to this Default Directory.

**Platform (please complete the following information):**
 - OS: Windows 10
 - .NET Runtime: 6
 - .NET Framework: 4.8
 - MailKit Version: 4.0.0
 - .Net 6 Console Application

**Exception**
Exception thrown: 'MailKit.Security.AuthenticationException' in System.Private.CoreLib.dll
An exception of type 'MailKit.Security.AuthenticationException' occurred in System.Private.CoreLib.dll but was not handled in user code
Authentication failed.

There is the code I'm using:
```csharp
var options = new PublicClientApplicationOptions
{
    ClientId = "...",
    TenantId = "...",
    RedirectUri = "http://localhost"
};

var publicClientApplication = PublicClientApplicationBuilder
    .CreateWithApplicationOptions(options)
    .Build();

var scopes = new string[] {
    "email",
    "offline_access",
    "https://outlook.office.com/IMAP.AccessAsUser.All"
};

using (var client = new ImapClient(new ProtocolLogger("imap.log")))
{
    var authToken = await publicClientApplication.AcquireTokenInteractive(scopes).ExecuteAsync();
    var oauth2 = new SaslMechanismOAuth2(authToken.Account.Username, authToken.AccessToken);
    await client.ConnectAsync("outlook.office365.com", 993, SecureSocketOptions.SslOnConnect);
    await client.AuthenticateAsync(oauth2);
    var inbox = client.Inbox;
    inbox.Open(FolderAccess.ReadOnly);

    var messages = inbox.Search(SearchQuery.All);

    foreach (var uid in messages)
    {
        var message = inbox.GetMessage(uid);
        Console.WriteLine(message.Subject);
    }

    await client.DisconnectAsync(true);
}
```

**Protocol Logs**
`
Connected to imaps://outlook.office365.com:993/
S: * OK The Microsoft Exchange IMAP4 service is ready. [SwBMADEAUABSADAAMgBDAEEAMAAwADEAOAAuAGEAcABjAHAAcgBkADAAMgAuAHAAcgBvAGQALgBvAHUAdABsAG8AbwBrAC4AYwBvAG0A]
C: A00000000 CAPABILITY
S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
S: A00000000 OK CAPABILITY completed.
C: A00000001 AUTHENTICATE XOAUTH2 ********
S: A00000001 NO AUTHENTICATE failed.
`

**Azure AD Setting**
![Screenshot 2023-04-19 at 11 58 01 AM](https://user-images.githubusercontent.com/16193686/232963421-2ec4dba7-c3b4-4b28-bdac-2d9255eecb35.png)
![Screenshot 2023-04-19 at 11 57 49 AM](https://user-images.githubusercontent.com/16193686/232963441-19514347-5b06-43de-a16c-eda3a2e04794.png)

I'm new to MailKit. Thanks for your help!

**Comment by jstedfast** [comment]
Looks like you copied the wrong RedirectUri.

Try `RedirectUri = "https://login.microsoftonline.com/common/oauth2/nativeclient"` instead.

**Comment by wAikAp** [comment]
> Looks like you copied the wrong RedirectUri.
> 
> Try `RedirectUri = "https://login.microsoftonline.com/common/oauth2/nativeclient"` instead.

Thanks, I have found the reason, is not about the rediectUri. Is about the account and its organization.
When I register the App with my organization's admin account and then log in to another testing account in the organization, it succeeds.
At first I used ac1@fatherCompany.com to create App registration on Azure AD. Then I added testing@mycompany.net to this AD as Member. But I don't know why the authentication is not successful (Azure is set to "Accounts in any organizational directory (Any Azure AD directory - Multitenant)") can successfully obtain the Access token but the authentication fails. Then I use admin@mycompany.net to create another App registration on the organization's Azure AD and use the client Id and tenant Id of this App Then use testing@mycompany.net to successfully log in and pass the authentication. From this, my inference is that the app under the same organization needs to be used to successfully pass the authentication?
